core/arch/arm/kernel/thread.c

```diff
@@ -553,6 +555,39 @@ void thread_handle_std_smc(struct thread_smc_args *args)
 		thread_alloc_and_run(args);
 }
 
+/* Helper routine for the assembly function thread_std_smc_entry() */
+void __thread_std_smc_entry(struct thread_smc_args *args)
+{
+	struct thread_ctx *thr = threads + thread_get_id();
+
+	if (!thr->rpc_arg) {
+		paddr_t parg;
+		void *arg;
+
+		parg = thread_rpc_alloc_arg(
+				TEESMC32_GET_ARG_SIZE(RPC_MAX_PARAMS));
+		if (!parg || !TEE_ALIGNMENT_IS_OK(parg, struct teesmc32_arg) ||
+		     core_pa2va(parg, &arg)) {
+			thread_rpc_free_arg(parg);
+			args->a0 = TEESMC_RETURN_ENOMEM;
+			return;
+		}
+
+		thr->rpc_arg = arg;
+		thr->rpc_parg = parg;
+	}
+
+	thread_std_smc_handler_ptr(args);
+
+	/*
+	 * Free the RPC argument to avoid having allocated shared memory
+	 * when all threads are done.
+	 */
+	thread_rpc_free_arg(thr->rpc_parg);
```

Comment:
Perhaps we should remove this step as it causes the allocation above to be performed on each entry.